File: lib/ftw_realtime/marketplace/conversation.ex
defmodule FtwRealtime.Marketplace.Conversation do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  schema "conversations" do
    belongs_to(:job, FtwRealtime.Marketplace.Job)
    belongs_to(:homeowner, FtwRealtime.Marketplace.User)
    belongs_to(:contractor, FtwRealtime.Marketplace.User)

    has_many(:messages, FtwRealtime.Marketplace.Message)

    timestamps(type: :utc_datetime)
  end

  def changeset(conversation, attrs) do
    conversation
    |> cast(attrs, [:job_id, :homeowner_id, :contractor_id])
    |> validate_required([:job_id, :homeowner_id, :contractor_id])
    |> foreign_key_constraint(:job_id)
    |> foreign_key_constraint(:homeowner_id)
    |> foreign_key_constraint(:contractor_id)
    |> unique_constraint([:job_id, :homeowner_id, :contractor_id])
  end
end

File: src/main/kotlin/com/strata/ftw/domain/entity/Conversation.kt
package com.strata.ftw.domain.entity

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.UUID

@Entity
@Table(
    name = "conversations",
    uniqueConstraints = [UniqueConstraint(columnNames = ["job_id", "homeowner_id", "contractor_id"])]
)
class Conversation(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "uuid")
    var id: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "job_id")
    var job: Job? = null,

    @Column(name = "job_id", insertable = false, updatable = false)
    var jobId: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "homeowner_id")
    var homeowner: User? = null,

    @Column(name = "homeowner_id", insertable = false, updatable = false)
    var homeownerId: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contractor_id")
    var contractor: User? = null,

    @Column(name = "contractor_id", insertable = false, updatable = false)
    var contractorId: UUID? = null,

    @OneToMany(mappedBy = "conversation", fetch = FetchType.LAZY)
    var messages: MutableList<Message> = mutableListOf(),

    @CreationTimestamp
    @Column(name = "inserted_at", updatable = false)
    var insertedAt: Instant? = null,

    @UpdateTimestamp
    @Column(name = "updated_at")
    var updatedAt: Instant? = null
)
